pandas/core/internals/blocks.py

```diff
@@ -434,40 +434,56 @@ def fillna(
         fillna on the block with the value. If we fail, then convert to
         ObjectBlock and try again
         """
+        # TODO: Handle inf_as_na, we need to get option and pass to cython funcs
         inplace = validate_bool_kwarg(inplace, "inplace")
-
-        mask = isna(self.values)
-        mask, noop = validate_putmask(self.values, mask)
-
-        if limit is not None:
-            limit = libalgos.validate_limit(None, limit=limit)
-            mask[mask.cumsum(self.ndim - 1) > limit] = False
+        arr = self if inplace else self.copy()
+        limit = libalgos.validate_limit(
+            len(self) if self.ndim == 1 else self.shape[1], limit=limit
+        )
 
         if not self._can_hold_na:
-            if inplace:
-                return [self]
+            return [arr]
+
+        if not self.is_extension:
+            if self._can_hold_element(value):
+                if self.ndim == 1:
+                    if is_list_like(value):
+                        # TODO: Verify EA case
+                        if is_extension_array_dtype(value):
+                            mask = value.isna()
+                            value = np.asarray(value[mask], dtype=object)
```

Comment:
going straight to object may be unnecessarily heavy for e.g. Categorical `value`.